Source/Details/ASCollectionViewLayout.m

```diff
@@ -0,0 +1,162 @@
+//
+//  ASCollectionViewLayout.m
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <AsyncDisplayKit/ASCollectionViewLayout.h>
+
+#import <AsyncDisplayKit/ASCellNode+Internal.h>
+
+#import <AsyncDisplayKit/ASAssert.h>
+#import <AsyncDisplayKit/ASCollectionElement.h>
+#import <AsyncDisplayKit/ASCollectionView.h>
+#import <AsyncDisplayKit/ASCollectionViewFlowLayoutCalculator.h>
+#import <AsyncDisplayKit/ASElementMap.h>
+#import <AsyncDisplayKit/ASLayout.h>
+
+@interface ASCollectionViewLayout () {
+  ASElementMap * __weak _elementMap;
+  CGSize _contentSize;
+  NSMapTable<ASCollectionElement *, UICollectionViewLayoutAttributes *> *_elementToLayoutAttributesMap;
+  BOOL _collectionViewIsPendingLayoutUpdate;
+}
+
+@end
+
+@implementation ASCollectionViewLayout
+
+- (instancetype)initWithLayoutCalculator:(id<ASCollectionViewLayoutCalculating>)calculator
+{
+  self = [super init];
+  if (self) {
+    _calculator = calculator;
+    _elementToLayoutAttributesMap = [NSMapTable mapTableWithKeyOptions:(NSMapTableStrongMemory | NSMapTableObjectPointerPersonality) valueOptions:NSMapTableCopyIn];
```

Comment:
Would be great to have a comment explaining the options as they affect this use case -- in particular CopyIn, which is not common to see in my experience.